libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -828,6 +828,28 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				StorageByteArrayElement(m_context).storeValue(*type, _functionCall.location(), true);
 			break;
 		}
+		case FunctionType::Kind::ByteArrayPop:
+		case FunctionType::Kind::ArrayPop:
+		{
+			_functionCall.expression().accept(*this);
+			solAssert(function.parameterTypes().size() == 0, "");
+
+			// TODO: Extract proper type for regular arrays
+			shared_ptr<ArrayType> arrayType = make_shared<ArrayType>(DataLocation::Storage);
```

Comment:
This is not the correct type. This is used as `x.pop()` and the type you want to have is the type of `x`, so you have to climb up the AST with something like `dynamic_cast<MemberAccess const&>(_functionCall.expression()).expression().type()`. Since there might be some edge cases like `var y = x.pop; y()`, where the AST does not have this form, you have to check that the dynamic cast does not fail. This means we have to add an appropriate check in `TypeChecker::visit(FunctionCall`.